aws/s3_test.go

```diff
@@ -132,332 +161,338 @@ func (b TestS3Bucket) addObject(s SetupInfo, fileName string, fileBody string) e
-}
-
-func TestNuke_EmptyS3Bucket(t *testing.T) {
-	s := setupNukeTests(t)
-
-	// Create test bucket
-	bucket := TestS3Bucket{
-		name: s.bucketName,
-	}
-
-	err := bucket.create(s.svc)
-	if err != nil {
-		assert.Failf(t, "Failed to create test bucket", errors.WithStackTrace(err).Error())
 	}
-
-	// Please note that we are passing the same session that was used to create the bucket
-	// This is required so that the defer cleanup call always gets the right bucket region
-	// to delete
-	defer nukeAllS3Buckets(s.awsSession, []*string{aws.String(s.bucketName)}, 1000)
-
-	// Create separate session object as bucket region may not be equal to session region
-	awsSession, err := createSession("")
-	if err != nil {
-		assert.Failf(t, "Failed to create random session", errors.WithStackTrace(err).Error())
+	for _, tc := range testCases {
+		testFunc, err := genTestListS3Bucket(tc.args)
+		if err != nil {
+			assert.Fail(t, errors.WithStackTrace(err).Error())
+		}
+		t.Run(tc.name, testFunc)
 	}
+}
 
-	// Nuke the test bucket
-	_, err = nukeAllS3Buckets(s.awsSession, []*string{aws.String(s.bucketName)}, 1000)
-	if err != nil {
-		assert.Fail(t, errors.WithStackTrace(err).Error())
-	}
+// TestNukeS3BucketArgs represents arguments for TestNukeS3Bucket
+type TestNukeS3BucketArgs struct {
+	isVersioned       bool
+	checkDeleteMarker bool
+	objectCount       int
+	objectBatchsize   int
+	shouldNuke        bool
+}
 
-	// Verify that - after nuking test bucket - it should not exist
-	bucketNamesPerRegion, err := getAllS3Buckets(awsSession, time.Now().Add(1*time.Hour), []string{s.region}, s.bucketName, 100)
+// genTestNukeS3Bucket - generates the test function for TestNukeS3Bucket
+func genTestNukeS3Bucket(args TestNukeS3BucketArgs) (func(t *testing.T), error) {
+	awsParams, err := newAWSParams()
 	if err != nil {
-		assert.Failf(t, "Failed to list S3 Buckets", errors.WithStackTrace(err).Error())
-	} else {
-		assert.NotContains(t, bucketNamesPerRegion[s.region], aws.String(s.bucketName))
+		return nil, fmt.Errorf("Failed to setup AWS params - %s", err.Error())
 	}
-}
-
-func testNukeS3BucketWrapper(t *testing.T, args *TestNukeS3BucketArgs, shouldNuke bool) {
-	s := setupNukeTests(t)
 
 	// Create test bucket
 	bucket := TestS3Bucket{
-		name:        s.bucketName,
+		name:        genTestBucketName(),
 		isVersioned: args.isVersioned,
 	}
-	err := bucket.create(s.svc)
+	err = bucket.create(awsParams.svc)
 	if err != nil {
-		assert.Failf(t, "Failed to create test bucket", errors.WithStackTrace(err).Error())
+		return nil, fmt.Errorf("Failed to create test bucket - %s", err.Error())
 	}
 
 	awsSession, err := createSession("")
 	if err != nil {
-		assert.Failf(t, "Failed to create random session", errors.WithStackTrace(err).Error())
+		return nil, fmt.Errorf("Failed to create random session - %s", err.Error())
 	}
 
-	defer nukeAllS3Buckets(s.awsSession, []*string{aws.String(s.bucketName)}, 1000)
+	if args.objectCount > 0 {
```

Comment:
This routine to setup the bucket should happen in the subtest function so that it can be done in parallel. Otherwise, we will be serially constructing each bucket for each test which will be slow.

Reply:
Also, the error handling is a bit tricky in this style as the deletion happens in the subtest. Ideally, we want to always run the cleanup function if we get far enough to create the bucket. This style makes it tricky to implement that because we might be incentivized to cut the test short if this function returns an error, but the cleanup function is in the subtest function.

Reply:
This should be resolved as per the latest commit.

Reply:
Thanks for resolving this!